// @mintlify/cli/src/scrape.tsx
import { MintConfigType } from '@mintlify/models';
import { addLog, ErrorLog, SuccessLog, SpinnerLog, InfoLog } from '@mintlify/previewing';
import {
  scrapePageGroup,
  scrapeAllSiteTabs,
  htmlToHast,
  detectFramework,
  framework,
  fetchPageHtml,
  write,
  getErrorMessage,
  generateOpenApiPages,
  FINAL_SUCCESS_MESSAGE,
} from '@mintlify/scraping';
import { upgradeToDocsConfig } from '@mintlify/validation';

import { terminate } from './helpers.js';

export async function scrapeSite(url: string, filter?: string) {
  try {
    const urlObj = new URL(url);
    addLog(<SpinnerLog message={`Fetching ${urlObj.toString()}...`} />);
    const html = await fetchPageHtml(urlObj);
    addLog(<SuccessLog message={`Successfully retrieved HTML from ${urlObj.toString()}`} />);

    addLog(<SpinnerLog message="Scraping site..." />);
    const result = await scrapeAllSiteTabs(html, urlObj, { filter });
    if (result.success) {
      const mintConfig = result.data as MintConfigType;
      const docsConfig = upgradeToDocsConfig(mintConfig, {
        shouldUpgradeTheme: true,
      });
      docsConfig.theme = 'aspen';
      write('docs.json', JSON.stringify(docsConfig, undefined, 2));
      addLog(<SuccessLog message={FINAL_SUCCESS_MESSAGE} />);
    } else {
      addLog(<ErrorLog message={result.message} />);
      await terminate(1);
    }
    await terminate(0);
  } catch (error) {
    const errorMessage = getErrorMessage(error);
    addLog(<ErrorLog message={errorMessage} />);
    await terminate(1);
  }
}

export async function scrapePage(url: string) {
  try {
    const urlObj = new URL(url);
    addLog(<SpinnerLog message={`Fetching ${urlObj.toString()}...`} />);
    const html = await fetchPageHtml(urlObj);
    addLog(<SuccessLog message={`Successfully retrieved HTML from ${urlObj.toString()}`} />);

    const hast = htmlToHast(html);
    detectFramework(hast);

    const needsBrowser = framework.vendor === 'gitbook';
    addLog(<SpinnerLog message="Scraping page..." />);
    const results = await scrapePageGroup([urlObj], needsBrowser);
    const result = results[0] || {
      success: false,
      message: `An unknown error occurred when scraping ${url}`,
    };

    if (result.success) {
      addLog(
        <SuccessLog
          message={`Successfully scraped ${url} ${result.data ? `into ${result.data[1]}` : ''}`}
        />
      );
    } else {
      addLog(<ErrorLog message={result.message} />);
      await terminate(1);
    }
    await terminate(0);
  } catch (error) {
    const errorMessage = getErrorMessage(error);
    addLog(<ErrorLog message={errorMessage} />);
    await terminate(1);
  }
}

interface ScrapeOpenApiOptions {
  openapiLocation: string;
  writeFiles: boolean;
  outDir?: string;
  overwrite: boolean;
}

export async function scrapeOpenApi({
  openapiLocation,
  writeFiles,
  outDir,
  overwrite,
}: ScrapeOpenApiOptions) {
  try {
    addLog(<SpinnerLog message={`Processing OpenAPI spec from ${openapiLocation}...`} />);
    const { nav, isUrl } = await generateOpenApiPages(openapiLocation, {
      openApiFilePath: undefined,
      version: undefined,
      writeFiles,
      outDir,
      overwrite,
    });
    addLog(<SuccessLog message="Successfully generated OpenAPI pages" />);
    addLog(<InfoLog message="Navigation object suggestion:" />);
    addLog(<InfoLog message={JSON.stringify(nav, undefined, 2)} />);
    if (isUrl) {
      addLog(<InfoLog message="OpenAPI location suggestion:" />);
      addLog(<InfoLog message={`openapi: ${openapiLocation}`} />);
    }
    await terminate(0);
  } catch (error) {
    if (error instanceof Error) {
      addLog(<ErrorLog message={error.message} />);
    } else {
      addLog(<ErrorLog message={String(error)} />);
    }
    await terminate(1);
  }
}
