// @mintlify/cli/bin/workflow.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { jsx as _jsx } from "react/jsx-runtime";
import { select, input, editor } from '@inquirer/prompts';
import { addLog, addLogs, SuccessLog } from '@mintlify/previewing';
import fse from 'fs-extra';
import { Text } from 'ink';
import path from 'path';
import { CMD_EXEC_PATH } from './constants.js';
export function slugify(name) {
    return name
        .toLowerCase()
        .replace(/[^a-z0-9]+/g, '-')
        .replace(/^-|-$/g, '');
}
const CRON_FIELD = /^(\*|(\*\/\d+)|(\d+(-\d+)?(,\d+(-\d+)?)*))$/;
export function isValidCron(expr) {
    const fields = expr.trim().split(/\s+/);
    if (fields.length !== 5)
        return false;
    return fields.every((f) => CRON_FIELD.test(f));
}
function escapeYaml(value) {
    return value.replace(/\\/g, '\\\\').replace(/"/g, '\\"');
}
export function buildFrontmatter(config) {
    var _a;
    const lines = [];
    lines.push(`name: "${escapeYaml(config.name)}"`);
    if (config.triggerType === 'push') {
        lines.push('on:');
        lines.push('  push:');
        if (config.triggerRepos) {
            for (const repo of config.triggerRepos) {
                lines.push(`    - repo: "${escapeYaml(repo)}"`);
            }
        }
    }
    else {
        lines.push('on:');
        lines.push(`  cron: "${escapeYaml((_a = config.cronExpression) !== null && _a !== void 0 ? _a : '')}"`);
    }
    if (config.contextRepos && config.contextRepos.length > 0) {
        lines.push('context:');
        for (const repo of config.contextRepos) {
            lines.push(`  - repo: "${escapeYaml(repo)}"`);
        }
    }
    if (config.automerge) {
        lines.push('automerge: true');
    }
    return `---\n${lines.join('\n')}\n---`;
}
const sendUsageMessageForAI = () => {
    addLogs(_jsx(Text, { children: "Agent Detected - sending AI friendly prompt" }), _jsx(Text, { children: '<system-message>' }), _jsx(Text, { children: "Example usage of `mintlify workflow`. This command is interactive and creates a workflow file in `.mintlify/workflows/`." }), _jsx(Text, { children: "Workflow files are Markdown files with YAML frontmatter. Instead of running this command, you can directly create a `.mintlify/workflows/your-workflow.md` file." }), _jsx(Text, { children: "Frontmatter fields: name (string), on.cron (string) or on.push with repo list, context (array of objects with repo key), automerge (boolean, only when true)." }), _jsx(Text, { children: "The Markdown body contains the agent instructions/prompt." }), _jsx(Text, { children: '</system-message>' }));
};
export function addWorkflow() {
    return __awaiter(this, void 0, void 0, function* () {
        const docsJsonPath = path.join(CMD_EXEC_PATH, 'docs.json');
        if (!(yield fse.pathExists(docsJsonPath))) {
            throw new Error('docs.json not found in the current directory. Please run this command from your docs repository root.');
        }
        const isInteractive = process.stdin.isTTY;
        const isClaudeCode = process.env.CLAUDECODE === '1';
        const isAI = !isInteractive || isClaudeCode;
        if (isAI) {
            sendUsageMessageForAI();
            return;
        }
        const workflowName = yield input({
            message: 'Workflow name',
            default: 'Update changelog',
        });
        if (!workflowName.trim()) {
            throw new Error('Workflow name cannot be empty.');
        }
        const slug = slugify(workflowName);
        if (!slug) {
            throw new Error('Workflow name must contain at least one alphanumeric character.');
        }
        const triggerType = yield select({
            message: 'Trigger type',
            choices: [
                { name: 'Cron (scheduled)', value: 'cron' },
                { name: 'Push (on push to repo)', value: 'push' },
            ],
        });
        let cronExpression;
        let triggerRepos = [];
        if (triggerType === 'cron') {
            cronExpression = yield input({
                message: 'Cron expression',
                default: '0 9 * * 1',
                validate: (value) => isValidCron(value) ||
                    'Invalid cron expression. Expected 5 fields: minute hour day-of-month month day-of-week (e.g. 0 9 * * 1).',
            });
        }
        else {
            const triggerReposInput = yield input({
                message: 'Trigger repos (comma-separated, e.g. your-org/your-docs)',
                default: '',
            });
            triggerRepos = triggerReposInput
                .split(',')
                .map((r) => r.trim())
                .filter(Boolean);
        }
        const contextReposInput = yield input({
            message: 'Context repos (comma-separated, e.g. your-org/your-product, optional)',
            default: '',
        });
        const contextRepos = contextReposInput
            .split(',')
            .map((r) => r.trim())
            .filter(Boolean);
        const automergeChoice = yield select({
            message: 'Enable automerge?',
            choices: [
                { name: 'Yes', value: 'yes' },
                { name: 'No', value: 'no' },
            ],
        });
        const instructions = yield editor({
            message: 'Agent instructions (opens your default editor)',
            default: '# Agent Instructions\n\nDescribe what this workflow should do.\n\nFor example:\n- Update the changelog based on recent commits\n- Summarize recent PRs\n',
        });
        const frontmatter = buildFrontmatter({
            name: workflowName,
            triggerType,
            cronExpression,
            triggerRepos,
            contextRepos: contextRepos.length > 0 ? contextRepos : undefined,
            automerge: automergeChoice === 'yes',
        });
        const filename = slug + '.md';
        const workflowDir = path.join(CMD_EXEC_PATH, '.mintlify', 'workflows');
        const filePath = path.join(workflowDir, filename);
        const relativePath = path.relative(CMD_EXEC_PATH, filePath);
        yield fse.ensureDir(workflowDir);
        if (yield fse.pathExists(filePath)) {
            throw new Error(`A workflow already exists at ${relativePath}. Please choose a different name or delete the existing file.`);
        }
        yield fse.writeFile(filePath, frontmatter + '\n\n' + instructions.trim() + '\n');
        addLog(_jsx(SuccessLog, { message: `Workflow created at ${relativePath}` }));
    });
}
